### ISSUE TITLE
Growing a `List` with `add` method is memory inefficient

### ISSUE THREAD

**Issue by lucatrv** [dart-lang/sdk#50120]
I compared the following two sample codes, written in Dart and Rust:

Dart:
```dart
import 'dart:io';

void main() {
  var z = <double>[];
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }
  print(z.length);
  stdin.readLineSync();
}
```

Rust:
```rust
use std::io;

fn main() {
    let mut z = Vec::<f64>::new();
    for _ in 0..10000000 {
        z.push(0.0);
    }
    println!("{}", z.len());
    io::stdin().read_line(&mut String::new()).unwrap();
}
```

I compiled the Dart code with `dart compile exe test.dart` and the Rust code with `rustc test.rs` and `rustc -O test.rs` (the binary created by the latter command is optimized for performance, but memory wise it is equivalent).

Apart from the difference in performance between Dart and Rust, which is not so important to me, and understandable due for instance to the use of GC in Dart, what puzzles me is the memory use. While the Rust binary uses about 82 MB of memory, the Dart binary uses almost 269 MB.

I then tried to pre-allocate the full list at once, as follows:
```dart
import 'dart:io';

void main() {
  final z = List<double>.filled(10000000, 0.0);
  print(z.length);
  stdin.readLineSync();
}
```

and now the memory usage is down to less than 85 MB, in line with Rust.

Therefore it seems to me that the `List.add` method is highly memory inefficient. Maybe every time that a new value is added to the list, only a single element is allocated and added to the list? If I do not mistake the Rust compiler grows the vector only once every so many added values, by reserving several contiguous memory elements each time. Anyhow whatever is the reason, the Rust compiler seems to work much better for this use case.

Please notice that the above code is only for example, in my use case I do not know the final size of the list, so I cannot pre-allocate it at once like in the last Dart sample code.

I'm using Dart version 2.18.2 on Windows 10.

**Comment by mraleph** [comment]
The `List.add` does not grow element by element - it grows by doubling. 

The issue for higher memory usage is that GC needs to happen to reclaim older versions of the array. You can see this by doing `pmap (pidof test.exe)`:

```
$ pmap (pidof test.exe)
1117244:   /tmp/test.exe
...
00007f2df7100000 132100K rw---   [ anon ]
00007f2dff280000  65540K rw---   [ anon ]
00007f2e03300000  32772K rw---   [ anon ]
00007f2e05380000  16388K rw---   [ anon ]
00007f2e06400000   8196K rw---   [ anon ]
00007f2e06c80000   4100K rw---   [ anon ]
00007f2e07100000   2052K rw---   [ anon ]
00007f2e07380000   1028K rw---   [ anon ]
00007f2e07500000    516K rw---   [ anon ]
00007f2e07600000    260K rw---   [ anon ]
...
```

Notice all these memory regions sticking around - each twice the size of the previous one. These are backing stores for the array you are growing. Once GC happens all of them except one will disappear and the memory usage will go down. That's pretty typical for any GC managed language. Rust obviously does not have the same problem because it reclaims memory eagerly. 

I'd say there is nothing really do here. In general if you are building huge arrays like this I would not recommend building them element by element to begin with. This is rather wasteful. Either preallocate or use B-tree or similar data structures which splits huge array into smaller arrays.

/cc @rmacnak-google

**Comment by lucatrv** [comment]
@mraleph thank you for your very detailed explanation. Just to clarify, at first I also thought about a GC delay, but then I waited several seconds and nothing happened, so I thought that something else was happening. Now I tried again, and waited even more, and for a few minutes nothing happened, then the memory usage was reduced to about about 112 MB.
Is it normal for the GC to take so much time to kick in? Shouldn't it release memory once the program is inactive, or maybe after a few seconds?
I tried even using `sleep(Duration(seconds:60));` instead of `stdin.readLineSync();`, just in case it was fooling the GC, but memory was not released until the program exited...

**Comment by rmacnak-google** [comment]
`sleep(Duration(seconds:60))` and `stdin.readLineSync()` are both examples of synchronous foreign calls. From the VM's perspective, the program is not idle. The program must yield to the message loop for idle GC to be considered.

**Comment by mmcdon20** [comment]
@lucatrv try using `await Future.delayed(const Duration(seconds: 60));` instead.

**Comment by lucatrv** [comment]
> @lucatrv try using `await Future.delayed(const Duration(seconds: 60));` instead.

I tried with `await Future.delayed(const Duration(seconds: 60));` but the result is the same, memory usage stays at 269 MB until the program exits, does it behave differently to you?

**Comment by mraleph** [comment]
@lucatrv 

```dart
import 'dart:io';

var z = <double>[];

void main() async {
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }
  print(z.length);
  await Future.delayed(const Duration(seconds: 60));
}
```

The memory usage definitely drops to 128Mb. The idle notification actually triggers after 1ms of idle time and that causes Scavenge+MarkCompact:

```console
$ out/ProductX64/dart_precompiled_runtime --verbose-gc /tmp/test.dart.aot
vm-isolate: hard_threshold=8796093022208MB, soft_threshold=70MB, idle_threshold=1MB, reason=initial
main: hard_threshold=8796093022208MB, soft_threshold=70MB, idle_threshold=1MB, reason=initial
main: hard_threshold=8796093022208MB, soft_threshold=1MB, idle_threshold=2MB, reason=loaded
[              |                          |     |       |      | new gen     | new gen     | new gen | old gen       | old gen       | old gen     |  store  | delta used   ]
[ GC isolate   | space (reason)           | GC# | start | time | used (MB)   | capacity MB | external| used (MB)     | capacity (MB) | external MB |  buffer | new  | old   ]
[              |                          |     |  (s)  | (ms) |before| after|before| after| b4 |aftr| before| after | before| after |before| after| b4 |aftr| (MB) | (MB)  ]
[ main         ,  StartCMark(   old space),    1,   0.00,   0.3,   0.0,   0.0,   2.0,   2.0, 0.0, 0.0,    0.7,    0.7,    1.1,    1.1,   0.0,   0.0,   0,   0,   0.0,    0.0, ]
main: hard_threshold=8796093022208MB, soft_threshold=1MB, idle_threshold=2MB, reason=loaded
main: hard_threshold=8796093022208MB, soft_threshold=140MB, idle_threshold=2MB, reason=gc
[ main         ,   MarkSweep(    finalize),    2,   0.00,   0.2,   0.0,   0.0,   2.0,   2.0, 0.0, 0.0,    0.7,    0.7,    1.6,    1.6,   0.0,   0.0,   0,   0,   0.0,   -0.0, ]
10000000
[ main         ,    Scavenge(        idle),    3,   1.22,   0.2,   0.3,   0.0,   2.0,   2.0, 0.0, 0.0,  256.5,  256.5,  257.9,  257.9,   0.0,   0.0,   1,   0,  -0.3,    0.0, ]
main: hard_threshold=8796093022208MB, soft_threshold=268MB, idle_threshold=130MB, reason=gc
[ main         , MarkCompact(        idle),    4,   1.36, 140.6,   0.0,   0.0,   2.0,   2.0, 0.0, 0.0,  256.5,  128.6,  257.9,  129.0,   0.0,   0.0,   0,   1,   0.0, -127.8, ]
```

**Comment by lucatrv** [comment]
@mraleph I tested again your code on my system and memory stays at 269 MB. Actually this time I run the binary several times and noticed that a few times memory usage dropped to about 137 MB, but most of the times it did not. Not sure why this is happening, I'm on Windows 10 and I'm compiling the code with `dart compile exe test.dart`.
I see you're on Linux, have you tried under Windows? Tonight I can try under Linux myself.

**Comment by lucatrv** [comment]
@mraleph I tested the sample code with `await Future.delayed(const Duration(seconds: 60));` under Linux. Most of the times, memory was released shortly after running the binary. However a few times memory was not release, like under Windows.
Then I noticed that on Linux I have only 5GB of RAM, while on Windows I have 12 GB of RAM. I therefore tried to fill up the Windows memory by running other programs, and then run the binary again. In those conditions, most of the times memory was released very quickly and deeply, down to about 7 MB usage. My conclusion is that what affects memory release by the GC is the amount of system memory available, is this expected? I hope you can confirm this on your systems as well.

**Comment by mraleph** [comment]
@lucatrv how do you measure memory? Idle GC should happen irrespectively of amount of available memory.

**Comment by lucatrv** [comment]
@mraleph on Linux I use [`btop`](https://github.com/aristocratos/btop) (same as `top`), on Windows I use the Task Manager. Am I the only one experiencing this?...

**Comment by mraleph** [comment]
Can you try using `pmap` on the process ID on Linux? I see idle GC happening consistently and the memory usage drops just as expected.

**Comment by lucatrv** [comment]
Here is `pmap`'s output:

```
14673:   ./z.exe
000055b4f045e000   1792K r---- z.exe
000055b4f061e000   2896K r-x-- z.exe
000055b4f08f2000    208K r---- z.exe
000055b4f0926000      8K rw--- z.exe
000055b4f0928000   1780K rw---   [ anon ]
000055b4f0fb2000   9216K rw---   [ anon ]
00007f5f80380000 132100K rw---   [ anon ]
00007f5f88500000  65540K rw---   [ anon ]
00007f5f8c580000  32772K rw---   [ anon ]
00007f5f8e600000  16388K rw---   [ anon ]
00007f5f8f680000   8196K rw---   [ anon ]
00007f5f8ff00000   4100K rw---   [ anon ]
00007f5f90380000   2052K rw---   [ anon ]
00007f5f90600000   1028K rw---   [ anon ]
00007f5f90780000    516K rw---   [ anon ]
00007f5f90880000    260K rw---   [ anon ]
00007f5f908ff000      4K -----   [ anon ]
00007f5f90900000   1024K rw---   [ anon ]
00007f5f90afe000      4K -----   [ anon ]
00007f5f90aff000   1024K rw---   [ anon ]
00007f5f90bff000      4K -----   [ anon ]
00007f5f90c00000   2152K rw---   [ anon ]
00007f5f90e80000    512K rw---   [ anon ]
00007f5f90f00000    512K r----   [ anon ]
00007f5f90fbe000      4K -----   [ anon ]
00007f5f90fbf000   1024K rw---   [ anon ]
00007f5f910bf000    324K r---- z.exe
00007f5f91110000     12K rw---   [ anon ]
00007f5f91113000    340K r-x-- z.exe
00007f5f91168000     12K rw---   [ anon ]
00007f5f9116b000      4K rw--- z.exe
00007f5f9116c000      8K rw---   [ anon ]
00007f5f9116e000    136K r---- libc.so.6
00007f5f91190000   1384K r-x-- libc.so.6
00007f5f912ea000    348K r---- libc.so.6
00007f5f91341000     16K r---- libc.so.6
00007f5f91345000      8K rw--- libc.so.6
00007f5f91347000     52K rw---   [ anon ]
00007f5f91354000     56K r---- libm.so.6
00007f5f91362000    488K r-x-- libm.so.6
00007f5f913dc000    376K r---- libm.so.6
00007f5f9143a000      4K r---- libm.so.6
00007f5f9143b000      4K rw--- libm.so.6
00007f5f9143c000      4K r---- libpthread.so.0
00007f5f9143d000      4K r-x-- libpthread.so.0
00007f5f9143e000      4K r---- libpthread.so.0
00007f5f9143f000      4K r---- libpthread.so.0
00007f5f91440000      4K rw--- libpthread.so.0
00007f5f91441000      4K r---- libdl.so.2
00007f5f91442000      4K r-x-- libdl.so.2
00007f5f91443000      4K r---- libdl.so.2
00007f5f91444000      4K r---- libdl.so.2
00007f5f91445000      4K rw--- libdl.so.2
00007f5f91457000     64K rw---   [ anon ]
00007f5f91467000      4K r---- z.exe
00007f5f91468000      8K r---- z.exe
00007f5f9146a000      4K r---- z.exe
00007f5f9146b000      8K rw---   [ anon ]
00007f5f9146d000      4K r---- ld-linux-x86-64.so.2
00007f5f9146e000    152K r-x-- ld-linux-x86-64.so.2
00007f5f91494000     40K r---- ld-linux-x86-64.so.2
00007f5f9149e000      8K r---- ld-linux-x86-64.so.2
00007f5f914a0000      8K rw--- ld-linux-x86-64.so.2
00007fff891a9000    132K rw---   [ stack ]
00007fff891cb000     16K r----   [ anon ]
00007fff891cf000      8K r-x--   [ anon ]
ffffffffff600000      4K --x--   [ anon ]
 total           289188K
```

**Comment by mraleph** [comment]
@lucatrv what's the version of Dart on your Linux machine?

**Comment by lucatrv** [comment]
@mraleph on Linux I'm using the [package included in Arch Linux](https://archlinux.org/packages/community/x86_64/dart/).
On Windows the latest updated Flutter installation.
Have you tried on Windows with lots of free memory? Because on my system this behavior is consistent and reproducible.

**Comment by lucatrv** [comment]
@mraleph here are the outputs of `dart --version`:

On Linux:
```
Dart SDK version: 2.18.2 (stable) (Tue Sep 27 13:24:11 2022 +0200) on "linux_x64"
```

On Windows:
```
Dart SDK version: 2.18.2 (stable) (Tue Sep 27 13:24:11 2022 +0200) on "windows_x64"
```

Should I open a separate issue about this?

**Comment by mraleph** [comment]
I have tried 2.18.2 SDK (from an official archive, downloaded from [here](https://storage.googleapis.com/dart-archive/channels/stable/release/2.18.2/sdk/dartsdk-linux-x64-release.zip)) and it does not really behave like you report. 

I have compiled your code from https://github.com/dart-lang/sdk/issues/50120#issuecomment-1270697058. In this case `pmap` reports `total            27064K` - because `z` does not survive across `await`.
 
Are you sure that you are not running some stale binary or a different version of the SDK?

**Comment by lucatrv** [comment]
@mraleph I tried several times with 2.18.2 SDK on three different systems: one with Arch Linux, one with Windows 10, and one with Windows 11 (all updated to the latest releases). Results are very consistent, so I am now confident that this issue is not due to my system. Perhaps you and I are not performing the same steps, so I am going to describe them here as precisely as I can.

Save the following sample code in a file named `z.dart`:

```dart
Future<void> main() async {
  var z = <double>[];
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }
  await Future.delayed(const Duration(seconds: 60));
}
```

Compile the code with the following command: `dart compile exe z.dart`.

On Windows (steps on Linux are very similar), open the Task Manager and a terminal. Then run `z.exe` and watch the used memory. If most of the system memory is used by other apps, the memory used by process `z.exe` will quickly ramp up to about 268 MB and then descend to about 6-7 MB.
Instead, if lots of system memory is available (my laptop has 12 GB of RAM), then some times the memory reaches 268 MB and stays stable until the process is terminated. If the first time that you run `z.exe` this does not happen, please terminate the process with `Ctrl-C` and run it again. If you still do not see this happening, please go on terminating the process and running it again. It should be fairly easy to reproduce.

Please let me know if you can reproduce this behavior, and if I should open a separate issue about this... thanks

**Comment by mraleph** [comment]
@lucatrv I have tried now on Windows as well and I can't reproduce it. As I have mentioned before whether or not your computer has a lot of free memory or not can't influence this - Dart VM is not even looking at this data. The code in question is rather deterministic - you allocate your huge array, you reach `await`, idle notification is sent, GC happens, memory is reclaimed. The only possibility I can imagine is there is some sort of bug which sometimes causes us to miss idle notification - but without being able to reproduce it, I am more inclined to think that this is some artifact of measurement.

**Comment by lucatrv** [comment]
@mraleph can I send you a video showing the issue, or maybe could we have a short Zoom meeting so that I can show you what happens in real time?... I'm really puzzled because I have already verified this behavior on several systems...

**Comment by julemand101** [comment]
@lucatrv 
Can you try run the following version of the program and tell us the output you are getting? It will only work on Linux (maybe also MacOS but I have not tested that):
```dart
import 'dart:convert';
import 'dart:io';

Future<void> main() async {
  print('[${DateTime.now()}] Step 1  : ${getMemoryUsage()}');
  var z = <double>[];

  print('[${DateTime.now()}] Step 2  : ${getMemoryUsage()}');
  await Future<void>.delayed(const Duration(seconds: 5));

  print('[${DateTime.now()}] Step 3  : ${getMemoryUsage()}');
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }

  print('[${DateTime.now()}] Step 4  : ${getMemoryUsage()}');
  for (var i = 0; i < 5; i++) {
    await Future<void>.delayed(const Duration(seconds: 1));
    print('[${DateTime.now()}] Step 5-$i: ${getMemoryUsage()}');
  }
}

String getMemoryUsage() => const LineSplitter()
    .convert(Process.runSync('pmap', ['$pid']).stdout as String)
    .last;
```

I have tested this on Arch Linux (both on a laptop and a server) and a server running a very old version of Debian and I am getting this similar result across all platforms (both AOT and JIT) where memory are being freeed rather quickly:
```
[2022-10-30 11:49:21.254002] Step 1  :  total            28360K
[2022-10-30 11:49:21.259159] Step 2  :  total            28360K
[2022-10-30 11:49:26.264965] Step 3  :  total            27736K
[2022-10-30 11:49:26.616219] Step 4  :  total           258148K
[2022-10-30 11:49:27.628097] Step 5-0:  total            27224K
[2022-10-30 11:49:28.635677] Step 5-1:  total            27224K
[2022-10-30 11:49:29.645771] Step 5-2:  total            27224K
[2022-10-30 11:49:30.653701] Step 5-3:  total            27224K
[2022-10-30 11:49:31.661735] Step 5-4:  total            27224K
```

**Comment by julemand101** [comment]
Made a version that (I think...) also works on Windows:
```dart
import 'dart:convert';
import 'dart:io';

Future<void> main() async {
  print('[${DateTime.now()}] Step 1  : ${getMemoryUsage()}');
  var z = <double>[];

  print('[${DateTime.now()}] Step 2  : ${getMemoryUsage()}');
  await Future<void>.delayed(const Duration(seconds: 5));

  print('[${DateTime.now()}] Step 3  : ${getMemoryUsage()}');
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }

  print('[${DateTime.now()}] Step 4  : ${getMemoryUsage()}');
  for (var i = 0; i < 5; i++) {
    await Future<void>.delayed(const Duration(seconds: 1));
    print('[${DateTime.now()}] Step 5-$i: ${getMemoryUsage()}');
  }
}

String getMemoryUsage() =>
    Platform.isWindows ? getMemoryUsageWindows() : getMemoryUsageLinux();

String getMemoryUsageLinux() => const LineSplitter()
    .convert(Process.runSync('pmap', ['$pid']).stdout as String)
    .last;

String getMemoryUsageWindows() => '${int.parse(Process.runSync('powershell', [
          '-Command',
          '(Get-Process -PID $pid).WorkingSet64'
        ]).stdout as String) ~/ 1000}K';
```
Arch Linux JIT:
```
[2022-10-30 13:43:53.722011] Step 1  :  total           188756K
[2022-10-30 13:43:53.772891] Step 2  :  total           188756K
[2022-10-30 13:43:58.787285] Step 3  :  total           158348K
[2022-10-30 13:43:59.221910] Step 4  :  total           381084K
[2022-10-30 13:44:00.239823] Step 5-0:  total           157324K
[2022-10-30 13:44:01.246452] Step 5-1:  total           157324K
[2022-10-30 13:44:02.253429] Step 5-2:  total           157836K
[2022-10-30 13:44:03.260433] Step 5-3:  total           157836K
[2022-10-30 13:44:04.267452] Step 5-4:  total           157836K
```
Arch Linux AOT:
```
[2022-10-30 13:44:15.925482] Step 1  :  total            28368K
[2022-10-30 13:44:15.932518] Step 2  :  total            28368K
[2022-10-30 13:44:20.948453] Step 3  :  total            27744K
[2022-10-30 13:44:21.344581] Step 4  :  total           258156K
[2022-10-30 13:44:22.377806] Step 5-0:  total            27232K
[2022-10-30 13:44:23.383395] Step 5-1:  total            27232K
[2022-10-30 13:44:24.389400] Step 5-2:  total            27232K
[2022-10-30 13:44:25.395398] Step 5-3:  total            27232K
[2022-10-30 13:44:26.401397] Step 5-4:  total            27232K
```
Windows 11 JIT:
```
[2022-10-30 13:50:34.206689] Step 1  : 139464K
[2022-10-30 13:50:34.321621] Step 2  : 139788K
[2022-10-30 13:50:39.448103] Step 3  : 139239K
[2022-10-30 13:50:39.661455] Step 4  : 341164K
[2022-10-30 13:50:40.783480] Step 5-0: 340832K
[2022-10-30 13:50:41.907684] Step 5-1: 340848K
[2022-10-30 13:50:43.030489] Step 5-2: 340869K
[2022-10-30 13:50:44.146568] Step 5-3: 340885K
[2022-10-30 13:50:45.267429] Step 5-4: 340795K
```
Windows 11 AOT:
```
[2022-10-30 13:50:52.366508] Step 1  : 11943K
[2022-10-30 13:50:52.472268] Step 2  : 12021K
[2022-10-30 13:50:57.576853] Step 3  : 12529K
[2022-10-30 13:50:57.787200] Step 4  : 247734K
[2022-10-30 13:50:58.889185] Step 5-0: 247754K
[2022-10-30 13:51:00.010251] Step 5-1: 12210K
[2022-10-30 13:51:01.116019] Step 5-2: 12226K
[2022-10-30 13:51:02.240807] Step 5-3: 12242K
[2022-10-30 13:51:03.367962] Step 5-4: 12263K
```

An interesting observation is that Windows 11 JIT does seem to get into a situation where memory are not freed as quickly as the other tests. I changed the program to wait for 15 seconds and it would still not free the memory. I have not been able to reproduce this behavior on other platforms or in AOT.

**Comment by lucatrv** [comment]
@julemand101 I tested your code on my Windows 10 system, and was never able to reproduce the memory issue, neither compiling the code with `dart compile exe` nor running it directly with `dart run`.

Then started playing with your code by commenting more and more parts until I was able to reproduce the issue again. Finally I realized that if I run the following code (either with AOT or JIT compilation):

```dart
Future<void> main() async {
  var z = <double>[];
  for (int i = 0; i < 10000000; i++) {
    z.add(0.0);
  }
  for (var i = 0; i < 6; i++) {
    await Future.delayed(const Duration(seconds: 10));
  }
}
```

which is the same as the original one apart from the fact that instead of waiting a full 60 seconds stretch, it waits 6 stretches of 10 seconds each one, then memory stays at 268 MB for the first 10 seconds, and then drops to 6-7 MB afterwards. As if the GC was actually blocked while it is waiting for `Future.delayed(const Duration(seconds: 10))`. Again, this happens only if there is lots of free system memory, otherwise the used memory drops quickly to 6-7 MB.

Please let me know if you can reproduce this behavior...

**Comment by julemand101** [comment]
@lucatrv 
Tested your latest version of the program on a Windows 10 Pro machine with 64 GB memory (where about 29 GB was in use) and I cannot reproduce the behavior with your version of the program either with AOT or JIT. Memory are being freed quickly if I check the Detailed view Task Manager and look at "Memory (active private working set)".
